src/main/java/org/jitsi/videobridge/cc/vp8/VP8FrameProjection.java

```diff
@@ -0,0 +1,461 @@
+    {
+        long delta;
+        if (!vp8Frame.matchesSSRC(nextVP8Frame))
+        {
+            // this is a simulcast switch. The typical incremental value =
+            // 90kHz / 30 = 90,000Hz / 30 = 3000 per frame or per 33ms
+            delta = 3000 * Math.max(1, (nowMs - createdMs) / 33);
+        }
+        else
+        {
+            // compute and apply a delta
+            delta = RTPUtils.rtpTimestampDiff(
+                nextVP8Frame.getTimestamp(), vp8Frame.getTimestamp());
+        }
+
+        long nextTimestamp = timestamp + delta;
+        return nextTimestamp & RawPacket.TIMESTAMP_MASK;
+    }
+
+    /**
+     * Small utility method that computes and returns the starting sequence
+     * number to use in the projection of the frame that will be sent
+     * immediately after the {@link #vp8Frame}.
+     *
+     * @return the starting sequence number to use in the projection of the
+     * frame that will be sent immediately after the {@link #vp8Frame}.
+     */
+    private int nextStartingSequenceNumber()
+    {
+        int vp8FrameLength = RTPUtils.getSequenceNumberDelta(
+            vp8Frame.getMaxSequenceNumber(),
+            vp8Frame.getStartingSequenceNumber());
+
+        int nextStartingSequenceNumber
+            = startingSequenceNumber + vp8FrameLength + 1;
+
+        return nextStartingSequenceNumber & RawPacket.SEQUENCE_NUMBER_MASK;
+    }
+
+    /**
+     * Small utility method that computes and returns the extended picture id to
+     * use in the projection of the frame that will be sent immediately after
+     * the {@link #vp8Frame}.
+     *
+     * @return the extended picture id to use in the projection of the frame
+     * that will be sent immediately after the {@link #vp8Frame}.
+     */
+    private int nextExtendedPictureId()
+    {
+        return (extendedPictureId + 1)
+            & DePacketizer.VP8PayloadDescriptor.EXTENDED_PICTURE_ID_MASK;
+    }
+
+    /**
+     * Rewrites an RTP packet that belongs to {@link #vp8Frame}. If this is the
+     * first packet of a frame and if a packet cache has been specified, we
+     * piggy-backing any missed packets.
+     *
+     * @param cache the cache to pull piggy-backed packets from.
+     * @param rtpPacket the RTP packet to project.
+     */
+    RawPacket[] rewriteRtp(@NotNull RawPacket rtpPacket, RawPacketCache cache)
+    {
+        int originalSequenceNumber = rtpPacket.getSequenceNumber();
+
+        rewriteRtpInternal(rtpPacket);
+
+        if (originalSequenceNumber != vp8Frame.getStartingSequenceNumber())
+        {
+            return AdaptiveTrackProjection.EMPTY_PACKET_ARR;
+        }
+
+        // We piggy-back any re-ordered packets of this frame.
+        long vp8FrameSSRC = vp8Frame.getSSRCAsLong()
+            , vp8FrameTimestamp = vp8Frame.getTimestamp();
+
+        List<RawPacket> piggyBackedPackets = new ArrayList<>();
+        boolean done = cache == null;
+        while (!done)
```

Comment:
so, this is basically finding every sequence number after the packet we're processing (which we've detected as being the first packet of the frame) which is present in the cache, right?  at least, until we hit one of a different frame (which we check by comparing the timestamps)?

Reply:
Yeah, nicely put :)

Reply:
I included your description in the code, so future generations won't have to interpret the code in their heads, but rather be illuminated  by your words :)